crossbeam/algorithm/synthesis.py

```diff
@@ -69,10 +76,37 @@ def update_with_better_value(result_value, all_value_dict, all_values, model,
           old_value, old_value.expression(), old_value.get_weight()))
 
 
-def copy_operation_value(value, all_values, all_value_dict):
+def copy_operation_value(operation, value, all_values, all_value_dict):
   assert isinstance(value, value_module.OperationValue)
   arg_values = [all_values[all_value_dict[v]] for v in value.arg_values]
-  return value_module.OperationValue(value.values, value.operation, arg_values)
+  if not value.values:
+    return operation.apply(arg_values, value.arg_variables, value.free_variables)
+  else:
+    return value_module.OperationValue(value.values, value.operation, arg_values)
```

Comment:
pass in a (deep?) copy of `value.arg_variables` and `value.free_variables` when constructing the OperationValue
https://github.com/ellisk42/xlambda/blob/main/crossbeam/dsl/value.py#L159

Reply:
I added deepcopy but it still didn't trigger the assertion error when dealing with some op.apply with non-empty arg_variables. maybe there is still something wrong but just in case if you might see anything on your end?

Reply:
i meant on line 82 when calling the OperationValue constructor (the one i linked), it takes arg_variables and free_variables as keyword arguments.

Reply:
done